src/engine/GroupBy.cpp

```diff
@@ -1010,31 +1006,34 @@ void GroupBy::computeGroupByForHashMapOptimization(
 
   size_t blockSize = 65536;
 
+  // Initialize vectors
+  std::vector<std::vector<AverageAggregationData>> aggregationData;
+  aggregationData.resize(numAggregates);
+
   for (size_t i = 0; i < subresult.size(); i += blockSize) {
     evaluationContext._beginIndex = i;
     evaluationContext._endIndex = std::min(i + blockSize, subresult.size());
 
     auto currentBlockSize =
         evaluationContext._endIndex - evaluationContext._beginIndex;
 
-    // Perform HashMap lookup for all groups in current block
-    std::vector<std::array<AverageAggregationData, numAggregates>*> hashEntries;
-
     auto getId = [&subresult, &evaluationContext, &columnIndex](size_t j) {
       return subresult(evaluationContext._beginIndex + j, columnIndex);
     };
 
-    // Create elements in map
+    // Perform HashMap lookup once for all groups in current block
+    std::vector<size_t> hashEntries;
     for (size_t j = 0; j < currentBlockSize; ++j) {
       auto id = getId(j);
-      map.try_emplace(id);
-    }
+      auto entry = map.try_emplace(id);
```

Comment:
```suggestion
      auto [iterator, wasAdded] = map.try_emplace(id);
```
(The feature is called `structured binding` in case you want to look it up)